### ISSUE TITLE
dump() relies on Object.toString() for type detection, which is legacy and may break some apps

### ISSUE THREAD

**Issue by mwojnars** [nodeca/js-yaml#648]
I ran across this issue after modifying Object.toString() to get more verbose debug messages in my app. Suddenly, `yaml.dump()` stopped working with `YAMLException: unacceptable kind of an object to dump`. I checked the source code and discovered that, to my surprise, js-yaml applies `toString()` to an object in order to check its type, and then performs string comparisons like `type === '[object Object]'` (in function `writeNode()`). This looks like a very risky way to check a type of an object, esp. when the ECMA specification explicitly says that:

> The above definition of toString preserves compatibility for legacy code that uses toString as a test for those specific kinds of built-in objects. It does not provide a reliable type testing mechanism for other kinds of built-in or program defined objects. In addition, **programs can use @@toStringTag in ways that will invalidate the reliability of such legacy type tests.**

(source: https://tc39.es/ecma262/multipage/fundamental-objects.html#sec-object.prototype.tostring)

In short, the spec says that using `toString()` for type tests is a legacy and should *not* be relied upon. In my app, I observed that this approach may break the core functionality of js-yaml in a *very unexpected* way, and is a blocker for me (and supposedly for some other users, too) to using js-yaml.

**Comment by puzrin** [comment]
In JS, modifying prototype methods of base objects is like shutting in to leg. That's not problem of this package.

**Comment by mwojnars** [comment]
All polyfills rely on modyfing base objects, so it's not a rare thing, rather a standard and officially accepted practice in JS - unavoidable in some cases given how flawed the standard API tends to be (e.g., re toString()). Sure, the issue I raised is not critical: most users will not notice, and myself, I already switched to another package. However, staying with this legacy way of type testing will surely come back to you more often over time, so I would be careful saying this is not your problem at all.

**Comment by puzrin** [comment]
Polyfills only exdend missing methods, but not modify behaviour of existing. Also, this package is very stable and support legacy browsers many years, without problems. I respect your right to do anything you wish, but can't consider things you do as good practice (=> no need efforts for support). Sorry if disappointed.
